src/filemanager.cpp

```diff
@@ -0,0 +1,2301 @@
+        // Retry every 5 ms for 50 ms, then every 10 ms for 100 ms, then
+        // every 100 ms
+        proj_sleep_ms(i < 10 ? 5 : i < 20 ? 10 : 100);
+    }
+    char **pasResult = nullptr;
+    int nRows = 0;
+    int nCols = 0;
+    sqlite3_get_table(hDB_,
+                      "SELECT 1 FROM sqlite_master WHERE name = 'properties'",
+                      &pasResult, &nRows, &nCols, nullptr);
+    sqlite3_free_table(pasResult);
+    if (nRows == 0) {
+        if (!createDBStructure()) {
+            sqlite3_close(hDB_);
+            hDB_ = nullptr;
+            return;
+        }
+    }
+
+    if (getenv("PROJ_CHECK_CACHE_CONSISTENCY")) {
+        checkConsistency();
+    }
+}
+
+// ---------------------------------------------------------------------------
+
+static const char *cache_db_structure_sql =
+    "CREATE TABLE properties("
+    " url          TEXT PRIMARY KEY NOT NULL,"
+    " lastChecked  TIMESTAMP NOT NULL,"
+    " fileSize     INTEGER NOT NULL,"
+    " lastModified TEXT,"
+    " etag         TEXT"
+    ");"
+    "CREATE TABLE chunk_data("
+    " id        INTEGER PRIMARY KEY AUTOINCREMENT CHECK (id > 0),"
+    " data      BLOB NOT NULL"
+    ");"
+    "CREATE TABLE chunks("
+    " id        INTEGER PRIMARY KEY AUTOINCREMENT CHECK (id > 0),"
+    " url       TEXT NOT NULL,"
+    " offset    INTEGER NOT NULL,"
+    " data_id   INTEGER NOT NULL,"
+    " data_size INTEGER NOT NULL,"
+    " CONSTRAINT fk_chunks_url FOREIGN KEY (url) REFERENCES properties(url),"
+    " CONSTRAINT fk_chunks_data FOREIGN KEY (data_id) REFERENCES chunk_data(id)"
+    ");"
+    "CREATE INDEX idx_chunks ON chunks(url, offset);"
+    "CREATE TABLE linked_chunks("
+    " id        INTEGER PRIMARY KEY AUTOINCREMENT CHECK (id > 0),"
+    " chunk_id  INTEGER NOT NULL,"
+    " prev      INTEGER,"
+    " next      INTEGER,"
+    " CONSTRAINT fk_links_chunkid FOREIGN KEY (chunk_id) REFERENCES chunks(id),"
+    " CONSTRAINT fk_links_prev FOREIGN KEY (prev) REFERENCES linked_chunks(id),"
+    " CONSTRAINT fk_links_next FOREIGN KEY (next) REFERENCES linked_chunks(id)"
+    ");"
+    "CREATE INDEX idx_linked_chunks_chunk_id ON linked_chunks(chunk_id);"
+    "CREATE TABLE linked_chunks_head_tail("
+    "  head       INTEGER,"
+    "  tail       INTEGER,"
+    "  CONSTRAINT lht_head FOREIGN KEY (head) REFERENCES linked_chunks(id),"
+    "  CONSTRAINT lht_tail FOREIGN KEY (tail) REFERENCES linked_chunks(id)"
+    ");"
+    "INSERT INTO linked_chunks_head_tail VALUES (NULL, NULL);";
+
+bool DiskChunkCache::createDBStructure() {
+
+    pj_log(ctx_, PJ_LOG_TRACE, "Creating cache DB structure");
+    if (sqlite3_exec(hDB_, cache_db_structure_sql, nullptr, nullptr, nullptr) !=
+        SQLITE_OK) {
+        pj_log(ctx_, PJ_LOG_ERROR, "%s", sqlite3_errmsg(hDB_));
+        return false;
+    }
+    return true;
+}
+
+// ---------------------------------------------------------------------------
+
+#define INVALIDATED_SQL_LITERAL "'invalidated'"
```

Comment:
Why is this up here?

Reply:
it is used just in the below method (and other ones). Adding a comment